File: LeetCode-Python/MinimumAbsoluteDifferenceinBST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def getMinimumDifference(self, root: Optional[TreeNode]) -> int:
        # Düğümler arasındaki farkları bulmak için bir liste 
        vals = [] 
        # Ağacı sıralı dolaşan yardımcı fonksiyon 
        def inorder(node): 
            if node: 
                inorder(node.left) 
                vals.append(node.val) 
                inorder(node.right) 
        # Ağacı sıralı dolaşarak düğüm değerlerini vals listesine ekliyoruz 
        inorder(root) 
        # Minimum farkı bulmak için varsayılan büyük bir değerle başlıyoruz 
        min_diff = float('inf') 
        # Sıralı değerler arasındaki farkları hesaplayarak minimum olanı buluyoruz 
        for i in range(1, len(vals)): 
            min_diff = min(min_diff, vals[i] - vals[i - 1]) 
        return min_diff

File: LeetCode-Cpp/MinimumAbsoluteDifferenceinBST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int getMinimumDifference(TreeNode* root) {
        // Önceki düğüm değerini saklamak için bir değişken 
        int prev = -1; 
        // Minimum farkı saklamak için bir değişken 
        int minDiff = INT_MAX; 
        // Ağaçta dolaşan yardımcı fonksiyon 
        inorderTraversal(root, prev, minDiff); 
        return minDiff;
    }
private: 
    void inorderTraversal(TreeNode* node, int& prev, int& minDiff) { 
        if (!node) return; 
        // Sol alt ağacı dolaş 
        inorderTraversal(node->left, prev, minDiff); 
        // Eğer önceki düğüm değeri varsa, minimum farkı güncelle 
        if (prev != -1) { 
            minDiff = min(minDiff, node->val - prev); 
        } 
        // Şu anki düğüm değerini önceki düğüm değeri olarak ayarla 
        prev = node->val; 
        // Sağ alt ağacı dolaş 
        inorderTraversal(node->right, prev, minDiff); }
};
